### PR TITLE
File_Engine: added Compute JSON Write and Read methods

### LINKED ISSUE #129: File_Engine: add Compute Write and Read methods for quick json writing/reading
#### Description:

As title, requested to simplify workflows @pawelbaran

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes #129

### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/Eh96SGBibiBItHu2SmNb2-QBQw2le3Fj4lntIk4FCnoYMQ?e=j3Hrpq

### CODE REVIEW

File: File_Engine/Compute/ReadFromJsonFile.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            }
+
+            return new List<object>() { converted };
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        private static bool FromJson(string json, out object result)
```
Comment (FraserGreenroyd) [inline File_Engine/Compute/ReadFromJsonFile.cs:96]: Is there a reason why this isn't using the Serialiser Engine? Apologies, I've not been fully up to date with discussions so if this has already been discussed just point me there!

File: File_Engine/Convert/ToJsonArray.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace BH.Engine.Adapters.File
+{
+    public static partial class Convert
+    {
+        /***************************************************/
+        /*** Methods                                     ***/
+        /***************************************************/
+
+        [Description("Serializes the objects and returns a json array.")]
```
Comment (FraserGreenroyd) [inline File_Engine/Convert/ToJsonArray.cs:43]:
```suggestion
        [Description("Serialises the objects and returns a json array.")]
```

File: File_Engine/File_Engine.csproj [hunk 3]
```diff
@@ -31,6 +31,9 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Adapter_oM">
+      <HintPath>..\..\..\..\..\ProgramData\BHoM\Assemblies\Adapter_oM.dll</HintPath>
```
Comment (FraserGreenroyd) [inline File_Engine/File_Engine.csproj:35]:
```suggestion
      <HintPath>C:\ProgramData\BHoM\Assemblies\Adapter_oM.dll</HintPath>
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

File: File_Engine/Compute/ReadFromJsonFile.cs (hunk 1 above)
Comment (pawelbaran) [reply File_Engine/Compute/ReadFromJsonFile.cs:96]: Up 👍

**Review by pawelbaran — CHANGES_REQUESTED** [review]
Looks good in general, however I have one concern on top of @FraserGreenroyd's comments: I think the file extension constraint requiring .json is too strict:
- it will not allow to save `ISettings` to .cfg file
- it may not allow valid files with uppercase extension (easy to fix with `ToLower` method)

**Comment by alelom** [comment]
> it may not allow valid files with uppercase extension (easy to fix with ToLower method)

Added `ToLower()` in both checks of the extension.

> it will not allow to save ISettings to .cfg file

[Cfg files have their own syntax depending on context](https://docs.fileformat.com/system/cfg/) and require a different treatment. Happy to discuss. I would consider that a completely separate feature from this. This still allows to write `settings.json` files a la VSCode.

File: File_Engine/Convert/ToJsonArray.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    public static partial class Convert
+    {
+        /***************************************************/
+        /*** Methods                                     ***/
+        /***************************************************/
+
+        [Description("Serialises the objects and returns a json array.")]
+        [Input("objects", "Objects you want to serialize and concatenate into a single JSON array.")]
+        [Output("jsonArray", "A JSON array with the serialized objects.")]
+        public static string ToJsonArray(this List<object> objects)
```
Comment (FraserGreenroyd) [inline File_Engine/Convert/ToJsonArray.cs:46]: What does this do differently from `BH.Engine.Serialiser.Convert.ToJsonArray`? Or is this also a left over of previous things?

File: File_Engine/Compute/WriteToJsonFile.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+            // If the file exists already, stop execution if `replace` is not true.
+            bool fileExisted = System.IO.File.Exists(filePath);
+            if (replace == false && fileExisted)
+            {
+                BH.Engine.Base.Compute.RecordWarning($"The file `{filePath}` exists already. To replace its content, set `{nameof(replace)}` to true.");
+                return false;
+            }
+
+            // Serialize to json and create the file and directory.
```
Comment (FraserGreenroyd) [inline File_Engine/Compute/WriteToJsonFile.cs:66]:
```suggestion
            // Serialise to json and create the file and directory.
```

File: File_Engine/Compute/WriteToJsonFile.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            fileInfo.Directory.Create(); // If the directory already exists, this method does nothing.
+
+            try
+            {
+                System.IO.File.WriteAllText(filePath, json);
+            }
+            catch (Exception e)
+            {
+                BH.Engine.Base.Compute.RecordError($"Error writing to file:\n\t{e.ToString()}");
+
```
Comment (FraserGreenroyd) [inline File_Engine/Compute/WriteToJsonFile.cs:78]:
```suggestion
```

File: File_Engine/Compute/WriteToJsonFile.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            // Make sure the .json extension is present.
+            if (!filePath.ToLower().EndsWith(".json") || string.IsNullOrWhiteSpace(filePath))
+            {
+                BH.Engine.Base.Compute.RecordError($"The filePath `{filePath}` must point to a JSON file. Make sure the file extension `.json` is present.");
+                return false;
+            }
+
+            // If the file exists already, stop execution if `replace` is not true.
+            bool fileExisted = System.IO.File.Exists(filePath);
+            if (replace == false && fileExisted)
```
Comment (FraserGreenroyd) [inline File_Engine/Compute/WriteToJsonFile.cs:60]:
```suggestion
            if (!replace && fileExisted)
```

File: File_Engine/Compute/WriteToJsonFile.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                return false;
+
+            if (objects == null)
+                return false;
+
+            // Make sure no invalid chars are present.
+            filePath = Path.Combine(Path.GetDirectoryName(filePath), Path.GetFileName(filePath));
+
+            // Make sure the .json extension is present.
+            if (!filePath.ToLower().EndsWith(".json") || string.IsNullOrWhiteSpace(filePath))
```
Comment (FraserGreenroyd) [inline File_Engine/Compute/WriteToJsonFile.cs:52]:
```suggestion
            if (string.IsNullOrWhiteSpace(filePath) || !filePath.ToLower().EndsWith(".json"))
```

It'll execute the first one first - so if the string is `null` then the `ToLower()` and `EndsWith()` methods might throw a `NullReferenceException`

File: File_Engine/Compute/WriteToJsonFile.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static bool WriteToJsonFile(List<object> objects, string filePath, bool replace = false, bool active = false)
+        {
+            if (!active)
+                return false;
+
+            if (objects == null)
+                return false;
```
Comment (FraserGreenroyd) [inline File_Engine/Compute/WriteToJsonFile.cs:46]:
```suggestion
            if (!active || objects == null)
                return false;
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
The CSProj change doesn't appear to have been pushed? The CSProj file will also need other compliance settings set, so might be easier to get BHoMBot to do it?

Still unsure why we have a `ToJsonArray` here as well as in Serialiser_Engine?

Also, documentation is missing for the new methods, but I'll run the bot's compliance to give a report of exactly what's needed :smile:

I agree with the discussion on Cfg being it's own problem - happy to discuss solutions for that that work for the use cases we have.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

File: File_Engine/Convert/ToJsonArray.cs (hunk 4 above)
Comment (alelom) [reply File_Engine/Convert/ToJsonArray.cs:46]: Good point. Kinda leftover from previous change where I forgot that the base Serialiser_Engine is able to treat Json Arrays now. However, this in addition allows to wrap non-bhomObjects and write primitive types, see example script, which the base serialiser can not do, and was requested by users.
Could be made a feature of the base Serialiser method. Could we agree on making this `private` in the meantime and keep it?

File: File_Engine/Compute/WriteToJsonFile.cs (hunk 8 above)
Comment (alelom) [reply File_Engine/Compute/WriteToJsonFile.cs]: This is now changed.

File: File_Engine/Convert/ToJsonArray.cs (hunk 4 above)
Comment (alelom) [reply File_Engine/Convert/ToJsonArray.cs]: I'm referring to this 
image

Comment (FraserGreenroyd) [reply File_Engine/Convert/ToJsonArray.cs]: Ok that makes sense, happy with this being `private` for now then and move a discussion on implementation into the Serialiser Engine for long  term 👍

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Review by FraserGreenroyd — APPROVED** [review]
This LGTM now, thanks @alelom 😄 will let @pawelbaran do a review too before merging

**Review by pawelbaran — APPROVED** [review]
LGTM 🔥

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling
@BHoMBot check core
@BHoMBot check installer
@BHoMBot check serialisation
@BHoMBot check versioning

**Pull request merged**